pipeline_dp/pipeline_backend.py

```diff
@@ -496,10 +552,70 @@ def __iter__(self) -> Iterator:
             self._cache = []
             for item in self._iterable:
                 self._cache.append(item)
-                yield item
             self._first_run_complete = True
+        yield from self._cache
+
+
+class LazySingleton:
+    """Represents a lazily evaluated object expected to resolve to a single value.
+
+    This class accepts either a list, which must contain exactly one element,
+    or an iterable, which is expected to yield exactly one element upon
+    iteration.
+
+    If initialized with a list, the single element is stored immediately.
+    If initialized with an iterable, the iterable is stored, and the element
+    is fetched and validated only when the `singleton()` method is first called.
+    The fetched element is then cached for subsequent calls.
+    """
+
+    def __init__(self, iterable_or_list: Union[List, Iterable]):
+        """Initializes the LazySingleton.
+
+        Args:
+            iterable_or_list: Either a list containing exactly one element,
+                              or an iterable expected to yield one element.
+
+        Raises:
+            ValueError: If the input is a list but does not contain exactly
+                        one element.
+            TypeError: If the input is neither a list nor an instance of
+                       collections.abc.Iterable.
+        """
+        self._singleton = None
+        self._iterable = None
+
+        if isinstance(iterable_or_list, list):
+            if len(iterable_or_list) != 1:
+                raise ValueError(
+                    f"Input list must contain exactly one element, but found "
+                    f"{len(iterable_or_list)} elements.")
+            self._singleton = iterable_or_list[0]
         else:
-            yield from self._cache
+            if not isinstance(iterable_or_list, Iterable):
+                raise TypeError(f"Input must be a list or an Iterable, but got "
+                                f"{type(iterable_or_list).__name__}.")
+            self._iterable = iterable_or_list
+
+    def singleton(self):
```

Comment:
should we make it a property?

Reply:
I'd say it's better a method, because it has non-trivial logic